File: go/read_planner.go
package turbodata

import "sort"

// Range is a single atomic byte range requested by the caller. After planning,
// each input range lives entirely inside exactly one ReadOp.
type Range struct {
	Offset int64
	Length int64
}

// ReadOp is one I/O request the fetcher will execute. It is the result of
// coalescing zero-or-more input ranges into a contiguous read (possibly
// reading wasted bytes between them) and optionally splitting an
// over-threshold coalesced range at internal range boundaries.
type ReadOp struct {
	Offset int64
	Length int64
}

// RangeLocation describes where a single original input range ended up after
// planning. The caller uses these to assemble a LoadedBytes once the fetcher
// has returned buffers.
type RangeLocation struct {
	OpIndex int // index into the returned []ReadOp
	InOpOff int // byte offset within that op's buffer
	Length  int
}

// Plan groups ranges into ReadOps according to s. The returned locations slice
// is parallel to the input ranges: locations[i] tells where ranges[i] lives in
// the produced ops.
//
// Algorithm:
//  1. Sort ranges by Offset, remembering original indices.
//  2. Coalesce: merge adjacent ranges whose gap is < CoalesceGap.
//  3. Split: a merged op whose length exceeds SplitThreshold is sliced at
//     internal range boundaries via greedy packing. A single range larger
//     than SplitThreshold stays as one oversize op — atomic units are never
//     broken.
func Plan(ranges []Range, s ReadStrategy) (ops []ReadOp, locations []RangeLocation) {
	locations = make([]RangeLocation, len(ranges))
	if len(ranges) == 0 {
		return nil, locations
	}

	// Sort by offset while remembering original indices.
	sorted := make([]int, len(ranges))
	for i := range sorted {
		sorted[i] = i
	}
	sort.SliceStable(sorted, func(i, j int) bool {
		return ranges[sorted[i]].Offset < ranges[sorted[j]].Offset
	})

	// Coalesce pass: produce groups of original indices, each group becoming
	// a single contiguous read at this stage.
	type group struct {
		offset  int64
		end     int64 // exclusive
		members []int // indices into ranges, in offset order
	}
	groups := make([]group, 0, len(ranges))
	for _, idx := range sorted {
		r := ranges[idx]
		if len(groups) == 0 {
			groups = append(groups, group{
				offset:  r.Offset,
				end:     r.Offset + r.Length,
				members: []int{idx},
			})
			continue
		}
		cur := &groups[len(groups)-1]
		gap := r.Offset - cur.end
		if gap >= 0 && gap < s.CoalesceGap {
			if r.Offset+r.Length > cur.end {
				cur.end = r.Offset + r.Length
			}
			cur.members = append(cur.members, idx)
			continue
		}
		groups = append(groups, group{
			offset:  r.Offset,
			end:     r.Offset + r.Length,
			members: []int{idx},
		})
	}

	// Split pass: for each group, emit one or more ReadOps. If
	// SplitThreshold is non-positive or larger than int64's range, the
	// whole group becomes one op.
	threshold := s.SplitThreshold
	splitDisabled := threshold <= 0

	for _, g := range groups {
		if splitDisabled || (g.end-g.offset) <= threshold {
			// Single op covers the whole group.
			opIndex := len(ops)
			ops = append(ops, ReadOp{Offset: g.offset, Length: g.end - g.offset})
			for _, mi := range g.members {
				locations[mi] = RangeLocation{
					OpIndex: opIndex,
					InOpOff: int(ranges[mi].Offset - g.offset),
					Length:  int(ranges[mi].Length),
				}
			}
			continue
		}

		// Greedy split at range boundaries.
		opStart := g.members[0]
		opStartIdx := 0
		opStartOffset := ranges[opStart].Offset
		opEndOffset := ranges[opStart].Offset + ranges[opStart].Length

		flushOp := func(lastMember int) {
			opIndex := len(ops)
			ops = append(ops, ReadOp{Offset: opStartOffset, Length: opEndOffset - opStartOffset})
			for k := opStartIdx; k <= lastMember; k++ {
				mi := g.members[k]
				locations[mi] = RangeLocation{
					OpIndex: opIndex,
					InOpOff: int(ranges[mi].Offset - opStartOffset),
					Length:  int(ranges[mi].Length),
				}
			}
		}

		for k := 1; k < len(g.members); k++ {
			mi := g.members[k]
			nextEnd := ranges[mi].Offset + ranges[mi].Length
			if nextEnd-opStartOffset > threshold {
				// Including this member would push past threshold; close
				// the current op without it.
				flushOp(k - 1)
				opStartIdx = k
				opStartOffset = ranges[mi].Offset
				opEndOffset = nextEnd
				continue
			}
			opEndOffset = nextEnd
		}
		flushOp(len(g.members) - 1)
	}

	return ops, locations
}

File: ts/src/read_planner.ts
// Copyright 2026 Wanjia He
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// Read planner: groups Ranges into ReadOps according to a ReadStrategy.
// Mirrors go/read_planner.go.

import type { ReadStrategy } from "./read_strategy.js";

export interface Range {
  /** int64. Absolute file offset. */
  offset: bigint;
  /** int64. Length in bytes. */
  length: bigint;
}

export interface ReadOp {
  offset: bigint;
  length: bigint;
}

export interface RangeLocation {
  /** Index into the produced ReadOp[]. */
  opIndex: number;
  /** Byte offset within that op's buffer (must fit in number; ops are bounded). */
  inOpOff: number;
  /** Length of this original range, in bytes. */
  length: number;
}

interface CoalesceGroup {
  offset: bigint;
  end: bigint; // exclusive
  members: number[]; // indices into input ranges, in offset order
}

/**
 * Plan groups Ranges into ReadOps.
 *
 * Precondition: ranges must be in non-decreasing offset order. All call sites
 * in this package satisfy this by construction (writer lays out groups and
 * chunks at monotonically increasing file offsets).
 *
 * Algorithm:
 *  1. Coalesce: merge adjacent ranges whose gap is < coalesceGap.
 *  2. Split: a coalesced op whose length exceeds splitThreshold is sliced at
 *     internal range boundaries via greedy packing. A single range larger
 *     than splitThreshold stays as one oversize op (atomic units never break).
 */
export function plan(
  ranges: Range[],
  s: ReadStrategy,
): { ops: ReadOp[]; locations: RangeLocation[] } {
  const locations: RangeLocation[] = new Array(ranges.length);
  if (ranges.length === 0) {
    return { ops: [], locations };
  }

  const groups: CoalesceGroup[] = [];
  for (let idx = 0; idx < ranges.length; idx++) {
    const r = ranges[idx]!;
    if (groups.length === 0) {
      groups.push({
        offset: r.offset,
        end: r.offset + r.length,
        members: [idx],
      });
      continue;
    }
    const cur = groups[groups.length - 1]!;
    const gap = r.offset - cur.end;
    if (gap >= 0n && gap < s.coalesceGap) {
      const newEnd = r.offset + r.length;
      if (newEnd > cur.end) {
        cur.end = newEnd;
      }
      cur.members.push(idx);
      continue;
    }
    groups.push({
      offset: r.offset,
      end: r.offset + r.length,
      members: [idx],
    });
  }

  const ops: ReadOp[] = [];
  const threshold = s.splitThreshold;
  const splitDisabled = threshold <= 0n;

  for (const g of groups) {
    if (splitDisabled || g.end - g.offset <= threshold) {
      const opIndex = ops.length;
      ops.push({ offset: g.offset, length: g.end - g.offset });
      for (const mi of g.members) {
        const r = ranges[mi]!;
        locations[mi] = {
          opIndex,
          inOpOff: Number(r.offset - g.offset),
          length: Number(r.length),
        };
      }
      continue;
    }

    // Greedy split at range boundaries.
    let opStartIdx = 0;
    let opStartOffset = ranges[g.members[0]!]!.offset;
    let opEndOffset =
      ranges[g.members[0]!]!.offset + ranges[g.members[0]!]!.length;

    const flushOp = (lastMember: number): void => {
      const opIndex = ops.length;
      ops.push({ offset: opStartOffset, length: opEndOffset - opStartOffset });
      for (let k = opStartIdx; k <= lastMember; k++) {
        const mi = g.members[k]!;
        const r = ranges[mi]!;
        locations[mi] = {
          opIndex,
          inOpOff: Number(r.offset - opStartOffset),
          length: Number(r.length),
        };
      }
    };

    for (let k = 1; k < g.members.length; k++) {
      const mi = g.members[k]!;
      const r = ranges[mi]!;
      const nextEnd = r.offset + r.length;
      if (nextEnd - opStartOffset > threshold) {
        flushOp(k - 1);
        opStartIdx = k;
        opStartOffset = r.offset;
        opEndOffset = nextEnd;
        continue;
      }
      opEndOffset = nextEnd;
    }
    flushOp(g.members.length - 1);
  }

  return { ops, locations };
}
